criu/plugin: Restore libhsakmt shared memory files

dayatsin-amd opened the pull request:
Libhsakmt(thunk) uses a shared memory file in /dev/shm/hsakmt_shared_mem
and its semaphore in /dev/shm/hsakmt_shared_mem. Adding a check during
checkpoint to see if these two files exist. If they exist then the
plugin will try to restore them during restore.

dayatsin-amd committed:
criu/plugin: Add support for dumping and restoring queues

This adds the capabilities to Checkpoint and Restore user mode queues
that are associated with a typical ROCm process such as memory queue
descriptors, control stacks and compute unit masks etc.

Also, serialize queue information returned during dump ioctl and store
them into image file.

Signed-off-by: David Yat Sin <david.yatsin@amd.com>
Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

rajbhar committed:
criu/plugin: Optimize the proto image size

Userptr buffer objects are system memory that criu handles so no need to
allocate space for such buffer objects. We should allocate space only
for GTT and VRAM memory since that is what we should save in our kfd
image file. The contents for userptr buffer objects will be drained by
criu via vmsplice since those are paged system memory and saved as
regular pages-*.img files.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

rajbhar committed:
criu/plugin: dump debug logs selectively

Plugin should log extra debug data only with -DDEBUG.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

dayatsin-amd committed:
criu/plugin: Support larger memory footprints

Use fopen/fwrite/fread to be able to handle bigger files.

Try to open /dev/kfd earlier during restore, so we can fail earlier if
amdgpu driver is not loaded.

Memcopy data directly into bo_info_test structure during dump to reduce
number of memcpy.

Moved some variable declarations so that the variables are only valid
during the scope where they are used.

Signed-off-by: David Yat Sin <david.yatsin@amd.com>

rajbhar committed:
criu/plugin: optimization for large bar read

PROCPIDMEM interface has memory bandwidth bottleneck during checkpoint
so it should be used only when direct mmap/memcpy is not possible.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

rajbhar committed:
criu/plugin: Add initial documentation for ROCm support.

 - placeholder and initial documentation.
 - keep with CRIU main documentation

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>
Signed-off-by: David Yat Sin <david.yatsin@amd.com>

dayatsin-amd committed:
criu/plugin: Re-adjust doorbell offset for queues

The doorbell offset depends on the gpu_id. When restoring on a different
node, the gpu_id may be different, so we need to re-adjust the doorbell
offsets.

Signed-off-by: David Yat Sin <david.yatsin@amd.com>

rajbhar committed:
criu/plugin: Pytorch container with criu

Add a Dockerfile to setup CRIU with its dependencies in an official ROCm
pytorch container. Support for plugin will be added in a future patch.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

dayatsin-amd committed:
criu/plugin: Dump and restore events

Add support for dumping and restoring events during CRIU checkpoint and
restore. Events are used to send notifications to applications when
certain tasks are completed or exceptions have occured.

Signed-off-by: David Yat Sin <david.yatsin@amd.com>
Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

rajbhar committed:
criu/plugin: Support AMD ROCm Checkpoint Restore with KFD

To support Checkpoint Restore with AMDGPUs for ROCm workloads, introduce
a new plugin to assist CRIU with the help of AMD KFD kernel driver. This
initial commit just provides the basic framework to build up further
capabilities. Like CRIU, the amdgpu plugin also uses protobuf to serialize
and save the amdkfd data which is mostly VRAM contents with some metadata.
We generate a data file "kfd.<id>.img" during the dump stage. On restore
this file is read and extracted to re-create various types of buffer
objects that belonged to the previously checkpointed process. Upon
restore the mmap page offset within a device file might change so we use
the new hook to update and adjust the mmap offsets for newly created
target process. This is needed for sys_mmap call in pie restorer phase.

With the current implementation (amdgpu_plugin), we support:
 - Only compute workloads such (Non Gfx) are supported
 - Single GPU
 - GPU visible inside a container
 - AMD GPU Gfx 9 Family
 - Pytorch Benchmarks such as BERT Base

Signed-off-by: David Yat Sin <david.yatsin@amd.com>
Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

rajbhar committed:
criu/restore: Introduce restore late stage hook

During criu restore phase when a device is getting restored with the
help of a plugin, some device specific operations need to be delayed
until criu finalizes the VMA placements in address space of the target
process. But by the time criu finalizes this, its too late since pie
phase is over and control is back to criu master process.

This change allows an external trigger to each resuming task to check
whether it has a device specific operation pending such as issuing an
ioctl call. Since this is called from criu master process context,
supply the pid of the target process and give a chance to each plugin
registered to run device specific operation if the target pid is valid.

A consumer for this will be added in a seperate patch.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

rajbhar committed:
criu/plugin: Implement restore late hook for kfd

For restarting MMU notifiers and to kick-start the evicted queues that
were created during a previuos criu restore plugin operation, we need an
external trigger that runs outside the target process context i.e. from
within master restorer context for each target (restored) pid. This
external trigger is required because we need to wait till criu pie
finalizes the VMAs relocation within the target process's address space.

When kfd recieves this ioctl, it starts the userptr worker and resumes
the MMU notifiers.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

rajbhar committed:
criu/plugin: Dockerfile for AMD criu repo

This sets up the pytorch environment for BERT Transformers and also sets
up CRIU along with all its dependencies including amdgpu plugin for
supporting CR with AMDGPUs.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

dayatsin-amd committed:
criu/plugin: Implement system topology parsing

Parse local system topology in /sys/class/kfd/kfd/topology/nodes/ and
store properties for each gpu in the CRIU image files. The gpu
properties can then be used later during restore to make the process is
restored on gpu's with similar properties.

Signed-off-by: David Yat Sin <david.yatsin@amd.com>

dayatsin-amd committed:
criu/plugin: Remap GPUs on checkpoint restore

The device topology on the restore node can be different from the
topology on the checkpointed node. The GPUs on the restore node may
have different gpu_ids, minor number. or some GPUs may have different
properties as checkpointed node. During restore, the CRIU plugin
determines the target GPUs to avoid restore failures caused by trying
to restore a process on a gpu that is different.

Signed-off-by: David Yat Sin <david.yatsin@amd.com>

dayatsin-amd committed:
criu/plugin: Add parameters to override mapping

Add optional parameters to override default behavior during restore.
These parameters are passed in as environment variables before executing
CRIU.

List of parameters:

KFD_DESTINATION_GPUS - override target gpu's
KFD_FW_VER_CHECK - disable firmware version check
KFD_SDMA_FW_VER_CHECK - disable SDMA firmware version check
KFD_CACHES_COUNT_CHECK - disable caches count check
KFD_NUM_GWS_CHECK - disable num_gws check
KFD_VRAM_SIZE_CHECK - disable VRAM size check
KFD_IGNORE_NUMA - ignore NUMA regions for CPUs

Signed-off-by: David Yat Sin <david.yatsin@amd.com>

rajbhar committed:
criu/files: *RFC* Don't cache fd for amdgpu devices

Restore operation fails when we perform CR operation of multiple
independent kfd proceses because criu caches the ids for the device
files with same mnt_ids, inode pair. This might not be the optimal
solution but a unique id is required for each kfd process for a
successful restore. Decoded fdinfo files also confirm a duplicate id for
amdgpu devices and thus cache lookup during checkpoint prevents the
actual metadata for the device VMA write in the image files so on a
subsequent restore the desired id is not found and hence restore fails
even though the actual image file e.g. renderDXXX.img exits.

This can be tested with any simple kfd application that uses AMD ROCt
library and creates a VRAM buffer object. CR on such individual
processes works fine but when we invoke more than one process from a
shell script and perform CR on this shell script, the restore operation
fails without this change.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

fxkamd committed:
criu/plugin: Add whitepaper document

Signed-off-by: Felix Kuehling <Felix.Kuehling@amd.com>
Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

rst0git commented on plugins/Dockerfile:
```diff
@@ -0,0 +1,95 @@
+#RUN mkdir -p /home/criu_build_dir
+WORKDIR /root/criu_build_dir
+#RUN cd /home/criu_build_dir
+RUN git clone https://github.com/checkpoint-restore/criu.git
```
Since this Dockerfile is part of the repository it might be better include the local files instead of cloning of the repository from GitHub, for example with the [`COPY`](https://docs.docker.com/engine/reference/builder/#copy) instruction. This will ensure that any changes that have been made locally will be included in the container image.

rajbhar replied:
Thank you for the feedback. With the current approach, we can just distribute the Dockerfile and it sets up everything inside the container for us.  Our internal QA teams find it useful this way. The other  Dockerfile even lets us chose a branch. These were actually intended for anyone who wants to test our plugin on their hardware with proper GPUs.

rst0git commented on plugins/kfd_plugin.c:
```diff
@@ -0,0 +1,1917 @@
+	return 0;
+}
+
+/* Returns 0 if successfull or name did not match. -errno if fail to parse  */
```
*successful

rajbhar replied:
ACK.

rst0git commented on plugins/kfd_plugin.c:
```diff
@@ -0,0 +1,1917 @@
+	{
+		DevinfoEntry *entry = xmalloc(sizeof(DevinfoEntry));
+		if (!entry) {
+			pr_err("Failed to allocate botest\n");
```
*Failed to allocate entry

rajbhar replied:
ACK

rst0git commented on plugins/kfd_plugin.c:
```diff
@@ -0,0 +1,1917 @@
+	for (int i = 0; i < num_queues; i++) {
+		QEntry *q_entry = xmalloc(sizeof(QEntry));
+		if (!q_entry) {
+			pr_err("Failed to allocate botest\n");
```
*q_entry

rajbhar replied:
ACK

rst0git commented on plugins/kfd_plugin.c:
```diff
@@ -0,0 +1,1917 @@
+	e->num_of_events = num_events;
+	return 0;
+}
+int kfd_plugin_init(int stage)
```
empty line before `kfd_plugin_init()`?

rajbhar replied:
Agreed

rst0git commented on plugins/kfd_plugin.c:
```diff
@@ -0,0 +1,1917 @@
+	unsigned char *buf;
+	size_t len;
+
+	printf("kfd_plugin: Enter cr_plugin_dump_file()- ID = 0x%x\n", id);
```
s/printf/pr_debug/

rajbhar replied:
Sure, with the current implementation, its takes a while to Checkpoint the VRAM contents so this gives a visual indication each time the plugin is called e.g. for kfd fd or for each render node fds etc.  We shall remove it going forward.

rst0git commented on plugins/kfd_plugin.c:
```diff
@@ -0,0 +1,1917 @@
+
+	e = criu_kfd__unpack(NULL, filestat.st_size, buf);
+	if (e == NULL) {
+		pr_err("Unable to parse the KFD message %#x", id);
```
This log message is missing a newline character at the end.

rajbhar replied:
ACK

rajbhar commented on plugins/kfd_plugin.c:
```diff
@@ -0,0 +1,1917 @@
+
+		sprintf(path, "%s/%s", TOPOLOGY_PATH, dirent_system->d_name);
+		if (stat(path, &stbuf)) {
+			pr_info("Cannot to access %s\n", path);
```
ACK

rajbhar commented:
Our current repository is a clone of official CRIU repository and not a fork so the PR was generated against criu-dev from our cloned repo. We are considering forking from upstream criu repo and hopefully retaining the same name as this one.

rajbhar committed:
criu/plugin: Add build options for amdgpu plugin

 - Make ROCm amdgpu plugin a part of standard criu build, install and
uninstall process.

Signed-off-by: Rajneesh Bhardwaj <rajneesh.bhardwaj@amd.com>

dayatsin-amd committed:
criu/plugin: Add unit tests for GPU remapping

Adding unit tests for GPU remapping code when checkpointing and
restoring on different nodes with different topologies.

dayatsin-amd committed:
criu/plugin: Read BO contents in parallel

Implement multi-threaded code to read contents of each GPU VRAM BO's in
parallel in order to speed up dumping process when using multiple GPUs.

dayatsin-amd committed:
criu/plugin: Restore libhsakmt shared memory files

Libhsakmt(thunk) uses a shared memory file in /dev/shm/hsakmt_shared_mem
and its semaphore in /dev/shm/hsakmt_shared_mem. Adding a check during
checkpoint to see if these two files exist. If they exist then the
plugin will try to restore them during restore.

rajbhar commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
+			return -EINVAL;
+		}
+
+		shared_mem_ptr = mmap(0, shared_mem_size, PROT_WRITE, MAP_SHARED, fd, 0);
```
PROT_READ | PROT_WRITE

dayatsin-amd replied:
I am not reading to this memory. I am just writing

rajbhar replied:
Ok, I was thinking about the global shared table that thunk creates. Then you should munmap in both save and restore calls.

dayatsin-amd replied:
ACK for munmap

rajbhar commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
+
+	/* First 4 bytes of shared file is the magic */
+	memcpy(shared_mem_magic, shared_mem_ptr, sizeof(*shared_mem_magic));
+	pr_debug("amdgpu_plugin: Shared mem magic:0x%x\n", *shared_mem_magic);
```
lets use plugin_log_msg since thats compiled only with DDEBUG

dayatsin-amd replied:
ACK

rajbhar reviewed (changes requested):
looks good overall, some changes requested.

rajbhar commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
+		return 0;
+
+	if (!stat(HSAKMT_SHM_PATH, &st)) {
+		pr_debug("amdgpu_plugin: %s already exists\n", HSAKMT_SHM_PATH);
```
I think you forgot to change this one to plugin_log_msg

dayatsin-amd replied:
For this one, it would be better to print if the user calls criu with -vvv

rajbhar commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
+	sem = sem_open(HSAKMT_SEM, O_CREAT, 0666, 0);
+	if (sem == SEM_FAILED) {
+		pr_perror("Failed to create %s\n", HSAKMT_SEM);
+		return -EACCES;
```
It returns SEM_FAILED with set error so just return -1 as pr_perror takes care of showing actual error

dayatsin-amd replied:
This will not affect behavior. Caller just checks for non-zero.

rajbhar replied:
This is just a nit and I can take care of this while applying this PR.  @fxkamd if you don't have any objections to this commit, I think its good to get included in our staging.

fxkamd replied:
We're making up error codes in a number of places. If we wanted to return the real error code, we'd return -errno. But the caller ignores this anyway. So it really doesn't matter.

fxkamd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -621,6 +626,86 @@ void *dump_bo_contents(void *_thread_data)
+	shared_mem_ptr = mmap(0, st.st_size, PROT_READ, MAP_SHARED, fd, 0);
+	if (shared_mem_ptr == MAP_FAILED) {
+		pr_perror("amdgpu_plugin: Failed to mmap shared mem %s\n", HSAKMT_SHM);
+		return -EACCES;
```
You should close the fd here.

dayatsin-amd replied:
ACK

fxkamd commented on plugins/amdgpu/Makefile:
```diff
@@ -9,14 +9,15 @@ include $(__nmk_dir)msg.mk
 
 CC      		:= gcc
 PLUGIN_CFLAGS  		:= -g -Wall -Werror -D _GNU_SOURCE -shared -nostartfiles -fPIC
+PLUGIN_LDFLAGS		:= -lrt -lpthread
```
This seems unrelated to the shared memory changes. Should this be part of the commit that adds multi-threaded saving of VRAM?

dayatsin-amd replied:
the shm_* functions need -lrt, and the sem_* functions need -lpthread.
The multi-threaded saving of VRAM should need:
+PLUGIN_LDFLAGS		:= -lpthread

I will add that to the multi-threaded commit if it is needed.
